plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs

```diff
@@ -212,4 +212,5 @@ deriving via GenericExMemoryUsage [a] instance ExMemoryUsage a => ExMemoryUsage
 
 -- TODO: presumably in the long run we'll want to do something a bit more careful here
 -- (maybe for costing purposes depth would be a better measure than total size, for instance).
-deriving via GenericExMemoryUsage Data instance ExMemoryUsage Data
+instance ExMemoryUsage Data where
+    memoryUsage _ = 1
```

Comment:
Why is it 1? Because the memory is already allocated before the evaluation starts?

Reply:
Because this is a hack with me trying to see if it fixes the performance issue...

Reply:
okay, don't mind me!

Reply:
> So one issue was the `ExMemory` instance for `Data` which was using the generic one, which will, I think, traverse the whole structure. 

Yes, I think that's right.  There is an [issue](https://jira.iohk.io/browse/SCP-2265) to remove these.  I think ultimately we';re going to want to have built-in values carrying sizes around with them so that they can be queried when we want to run the costing function.

Reply:
Yeah. But what we don't want is O(1) pattern-matching being O(n) because it traverses the data structure to work out how big it is before returning "1"!

Reply:
> Yeah. But what we don't want is O(1) pattern-matching being O(n) because it traverses the data structure to work out how big it is before returning "1"!

Absolutely.  I think that more complicated built-in types would have to carry a "size" value (or maybe more than one) about with them and adjust it when the structure gets updated, so a list would really be something like `(list, int)` internally, with the second value being the length, incremented with every `cons`.